pkg/validator/cluster.go

```diff
@@ -312,13 +312,61 @@ func validateGeneralClusterChanges(
 	valErrors []string,
 ) []string {
 
-	if cr.Spec.AppID != prevCr.Spec.AppID {
-		appModifiedMsg := fmt.Sprintf(
-			modifiedProperty,
-			"app",
-		)
-		valErrors = append(valErrors, appModifiedMsg)
+	crApp, err := catalog.GetApp(cr)
+	if err != nil {
+		valErrors = append(valErrors, err.Error())
+	}
+
+	prevCrApp, err := catalog.GetApp(prevCr)
+	if err != nil {
+		valErrors = append(valErrors, err.Error())
 	}
+
+	if crApp != nil && prevCrApp != nil {
+
+		rbKey := cr.Name + cr.Namespace
+		rbInfo := shared.ClusterRollbackInfoMap[rbKey]
+
+		if prevCrApp.Spec.DistroID != crApp.Spec.DistroID {
+			appModifiedMsg := fmt.Sprintf(
+				invalidDistroId,
+				crApp.Spec.DistroID,
+				prevCrApp.Spec.DistroID,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
+		if prevCrApp.Spec.Version == crApp.Spec.Version {
+			appModifiedMsg := fmt.Sprintf(
+				versionIsNotModified,
+				crApp.Spec.DistroID,
+				crApp.Spec.Version,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
```

Comment:
Let's say the versions are in fact different though. It seems like we need to be doing a few other checks here:

* If we're in the middle of an upgrade already, make sure we're only rolling back to the previous kdapp (whose identity should be stored in our kdcluster status, as mentioned elsewhere).
* Make sure this request doesn't contain spec changes other than the app change (may need to check this somewhere else).
* If this is an upgrade, do we want to enforce that the new app version is semantically greater than the old?

Reply:
Also, there's other necessary checks although this function may not be the best place to do them.
* A kdcluster spec change that changes the app is not allowed to change anything else (e.g. member counts).
* If an upgrade is in progress, no spec changes are allowed except for rollback.